### ISSUE TITLE
Automated snapshot zero size on Azure storage

### ISSUE THREAD

**Issue by codybum** [allegroai/trains#203]
I am using Azure as backend storage, sklearn to generate models, and joblib to dump model output.  The model is dumped as follows:  

```
from joblib import dump, load

id = str(uuid.uuid1())
with open("model_" + id + ".pkl", 'wb') as mf:
        dump(best_random, mf)
```

The model saves correctly on the disk:
```
3497284 Sep  6 17:06 model.joblib
```

Trains correctly detects that joblib is saving the model: 

```
2020-09-06 17:12:10,308 - trains.storage - INFO - Starting upload: /tmp/.trains.upload_model_d53bycin.tmp => trainshepa-a-pred/randomforest-1599412317.3747468.48cacad3245e4af0a5d5ada8bc92b2a3/models/model_1900783c-f064-11ea-99da-5cb9019252e5.pkl
2020-09-06 17:12:10,399 - trains.Task - INFO - Completed model upload to azure://[name removed].blob.core.windows.net/trains/hepa-a-pred/randomforest-1599412317.3747468.48cacad3245e4af0a5d5ada8bc92b2a3/models/model_1900783c-f064-11ea-99da-5cb9019252e5.pkl
```
However, while the correct directory structure and filenames are being created in Azure, the file size is zero bytes:

![image](https://user-images.githubusercontent.com/6785316/92331382-28224b80-f044-11ea-8c4e-688a06d754b8.png)

I suppose this could be a container or other configuration issue, but no errors are generated during the process.

Any thoughts would be appreciated.

**Comment by bmartinn** [comment]
Hi @codybum 

Could you verify that the local model is not zero bytes ? (just making sure here :)

Assuming that we passed that test, maybe this is some azure credentials thing, create vs write ?

You can quickly test you credentials, let's assume you have a local file `/tmp/a_file.txt`
Let's check if we can upload it to the same bucket manually:
```python
from trains import StorageManager
StorageManager.upload_file(local_file='/tmp/a_file.txt', remote_url='azure://[name removed].blob.core.windows.net/trains/hepa-a-pred/tmp_delete_me/a_file.txt')
```

BTW:
you are definitely correct on not outputting an error, the odd thing is that it reported the uploaded as completed successfully...
Once we nail down the bug, I would also like to understand why `Trains` failed to report the upload error...

**Comment by codybum** [comment]
Hi @bmartinn 

The manual upload from /tmp, worked just fine.

![image](https://user-images.githubusercontent.com/6785316/92389949-01b5ec00-f0e8-11ea-84a9-aa03a79011de.png)

I also tried to exclusively lock the model file during the dump thinking trains might be grabbing the file before it was closed, but this didn't do anything.
 
I will dig around the code and see if I can find anything.

**Comment by codybum** [comment]
I found the issue, but I am not sure how to fix it.  When the following line is called in trains/binding/frameworks/__init__.py:  
```
shutil.copy(files[0], temp_file)
```
https://github.com/allegroai/trains/blob/aa44ba854fa91ec75e56b84219a89cfca36ec4ea/trains/binding/frameworks/__init__.py#L425

file[0] points to the file generated by the joblib dump.  At the time at the time of execution, file[0] is zero bytes, so nothing is copied to the temporary file.  If execution is allowed to continue past model upload, the file referenced by file[0] will be saved normally.

Basically, it looks like trains is trying to grab the dump before it is created.

**Comment by bmartinn** [comment]
@codybum nicely done! 

[create_output_model](https://github.com/allegroai/trains/blob/aa44ba854fa91ec75e56b84219a89cfca36ec4ea/trains/binding/frameworks/__init__.py#L294) is called after the joblib stored the dump (i.e. serially) so in theory the file should not be zero.
Basically what should happen is that the [patched joblib dump](https://github.com/allegroai/trains/blob/aa44ba854fa91ec75e56b84219a89cfca36ec4ea/trains/binding/joblib_bind.py#L82) is calling the original joblib dump and then it will upload the created file. For some reason it seems the file was not flushed yet, although we specifically [flush](https://github.com/allegroai/trains/blob/aa44ba854fa91ec75e56b84219a89cfca36ec4ea/trains/binding/joblib_bind.py#L106) it, if it is a file stream and not a string.

I'm not aware of any async dump option in joblib, What do you think?

**Comment by codybum** [comment]
I might not be looking at the right thing, but at least for me the original_fn (which I think is the original dump function) returns "None" at this point:

https://github.com/allegroai/trains/blob/aa44ba854fa91ec75e56b84219a89cfca36ec4ea/trains/binding/joblib_bind.py#L91

It is unclear how eventually the model is saved, but it appears at least at the point you would expect (https://gemfury.com/stream/python:scikit-learn/-/content/externals/joblib/numpy_pickle.py) a filename.

**Comment by bmartinn** [comment]
Hi @codybum I'm happy to say I found the problem.
Seems the NumpyPickler builds another compressed stream on top of the file stream, this means the `flush` is called on the compressed stream , which does not actually flush to disk.
We are working on a fix, it should not take long to release, I'll update here once it is available.

**Comment by codybum** [comment]
@bmartinn Great news!

**Comment by bmartinn** [comment]
Hi @codybum 
An RC with a fix is out, could you verify the issue is solved?
```
pip install trains==0.16.2rc0
```

**Comment by codybum** [comment]
Happy to confirm that things seem to be working just fine!

![image](https://user-images.githubusercontent.com/6785316/92735328-51462300-f347-11ea-8176-5b2697a4cd3a.png)
